# Import Data and exploration

In [1]:
import pandas as pd
# REMOVE THE (#) BELOW IF IT'S THE FIRST TIME RUNNING THIS TO GET THE DATA OR YOU WANT TO UPDATE THE DATA #
                        # THEN PUT THEM BACK TO AVOID LONG EXECUTION TIME #

url1 = 'https://data.insideairbnb.com/united-states/ny/new-york-city/2026-06-14/visualisations/neighbourhoods.csv'
pd.read_csv(url1).to_csv('data/neighbourhoods.csv',index=False)

url2 = 'https://data.insideairbnb.com/united-states/ny/new-york-city/2026-06-14/data/listings.csv.gz'
pd.read_csv(url2).to_csv('data/listings.csv.gz',index=False)

neighbourhoods = pd.read_csv("data/neighbourhoods.csv")
listings = pd.read_csv("data/listings.csv.gz")

In [2]:
pd.set_option('display.max_columns', None)
t = '---------------------------------------'

In [3]:
neighbourhoods.head()

,neighbourhood_group,neighbourhood
0,Bronx,Allerton
1,Bronx,Baychester
2,Bronx,Belmont
3,Bronx,Bronxdale
4,Bronx,Castle Hill


In [4]:
listings.head()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,host_url,host_profile_id,host_profile_url,host_name,host_since,hosts_time_as_user_years,hosts_time_as_user_months,hosts_time_as_host_years,hosts_time_as_host_months,host_location,host_about,host_response_time,host_response_rate,host_acceptance_rate,host_is_superhost,host_thumbnail_url,host_picture_url,host_neighbourhood,host_listings_count,host_total_listings_count,host_verifications,host_has_profile_pic,host_identity_verified,neighbourhood,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,amenities,price,price_quote_checkin_date,price_quote_checkout_date,price_quote_total_price,price_quote_price_per_night,price_quote_raw,minimum_nights,maximum_nights,minimum_minimum_nights,maximum_minimum_nights,minimum_maximum_nights,maximum_maximum_nights,minimum_nights_avg_ntm,maximum_nights_avg_ntm,calendar_updated,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,availability_eoy,number_of_reviews_ly,estimated_occupancy_l365d,estimated_revenue_l365d,first_review,last_review,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,2539,https://www.airbnb.com/rooms/2539,20260614073253,2026-06-15,city scrape,11 Min to Manhattan • Prospect Park • Fast WiFi,"Bright, serene room in a renovated apartment h...",NaN,https://a0.muscache.com/pictures/hosting/Hosti...,2787,https://www.airbnb.com/users/show/2787,1.462506e+18,https://www.airbnb.com/users/profile/146250632...,John,NaN,17.0,9.0,15.0,5.0,"New York, NY",Educated professional living in Brooklyn. I l...,NaN,NaN,NaN,f,NaN,https://a0.muscache.com/im/pictures/user/User/...,NaN,5.0,NaN,NaN,t,t,NaN,Kensington,Brooklyn,40.645937,-73.972164,Private room in condo,Private room,2,1.0,1 shared bath,1.0,2.0,"[""Bed linens"", ""Hot water"", ""Exterior security...",$113.97,2026-07-11,2026-08-10,3419.19,113.97,"{""quote"": {""taxes"": null, ""currency"": ""USD"", ""...",30.0,730.0,30.0,30.0,730.0,730.0,30.0,730.0,NaN,t,4,34,64,339,2026-06-15,10,1,1,174,0,60,6838.0,2015-12-04,2026-05-28,4.80,4.67,4.90,4.89,5.00,4.78,4.78,NaN,NaN,5,0,5,0,0.08
1,6848,https://www.airbnb.com/rooms/6848,20260614073253,2026-06-14,city scrape,Only 2 stops to Manhattan studio,Comfortable studio apartment with super comfor...,NaN,https://a0.muscache.com/pictures/e4f031a7-f146...,15991,https://www.airbnb.com/users/show/15991,1.462507e+18,https://www.airbnb.com/users/profile/146250678...,Allen,NaN,17.0,1.0,15.0,5.0,"New York, NY",We love to travel. When we travel we like to s...,NaN,NaN,NaN,t,NaN,https://a0.muscache.com/im/users/15991/profile...,NaN,1.0,NaN,NaN,t,t,NaN,Williamsburg,Brooklyn,40.709350,-73.953420,Entire rental unit,Entire home/apt,3,1.0,1 bath,2.0,1.0,"[""Bed linens"", ""Hot water"", ""Fire extinguisher...",$117.27,2026-12-20,2027-01-19,3518.00,117.27,"{""quote"": {""taxes"": null, ""currency"": ""USD"", ""...",30.0,240.0,30.0,30.0,240.0,240.0,30.0,240.0,NaN,t,1,1,15,191,2026-06-14,198,3,0,27,4,180,21109.0,2009-05-25,2026-04-23,4.60,4.60,4.85,4.85,4.80,4.69,4.59,NaN,NaN,1,1,0,0,0.95
2,6872,https://www.airbnb.com/rooms/6872,20260614073253,2026-06-14,city scrape,Uptown Sanctuary w/ Private Bath (Month to Month),This charming distancing-friendly month-to-mon...,NaN,https://a0.muscache.com/pictures/miso/Hosting-...,16104,https://www.airbnb.com/users/show/16104,1.462507e+18,https://www.airbnb.com/users/profile/146250678...,Kae,NaN,17.0,1.0,15.0,5.0,"New York, NY",A former life in fashion and wellness h

In [5]:
print(listings.shape)
print(t)
print(listings.columns.tolist())
print(t)
print(listings.dtypes)
print(t)
print(listings.isna().sum().sort_values(ascending=False).head(20))
print(t)
print(listings[['price']].head(10))
print(t)
print(listings['price'].dtype)

(30259, 90)
---------------------------------------
['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name', 'description', 'neighborhood_overview', 'picture_url', 'host_id', 'host_url', 'host_profile_id', 'host_profile_url', 'host_name', 'host_since', 'hosts_time_as_user_years', 'hosts_time_as_user_months', 'hosts_time_as_host_years', 'hosts_time_as_host_months', 'host_location', 'host_about', 'host_response_time', 'host_response_rate', 'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url', 'host_picture_url', 'host_neighbourhood', 'host_listings_count', 'host_total_listings_count', 'host_verifications', 'host_has_profile_pic', 'host_identity_verified', 'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude', 'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms', 'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price', 'price_quote_checkin_date', 'price_quote_checkout_date', 'price_quote_total_price', 'price_quo

In [6]:
print(listings['neighbourhood_cleansed'].head())
print(t)
print(listings['neighbourhood_group_cleansed'].unique())
print(t)
print(neighbourhoods.shape)
print(t)
print(neighbourhoods.columns.tolist())
print(t)
print(neighbourhoods.head())


0      Kensington
1    Williamsburg
2     East Harlem
3     East Harlem
4     Fort Greene
Name: neighbourhood_cleansed, dtype: str
---------------------------------------
<ArrowStringArray>
['Brooklyn', 'Manhattan', 'Queens', 'Bronx', 'Staten Island']
Length: 5, dtype: str
---------------------------------------
(230, 2)
---------------------------------------
['neighbourhood_group', 'neighbourhood']
---------------------------------------
  neighbourhood_group neighbourhood
0               Bronx      Allerton
1               Bronx    Baychester
2               Bronx       Belmont
3               Bronx     Bronxdale
4               Bronx   Castle Hill


# Audit missing data
Report how many missing values exist in each column, and what percentage of rows are affected overall. Columns that are 100% empty should be identified separately from partially-missing ones.

PS: `neighbourhoods.csv` had no missing Values at the time this analysis was made

In [7]:
col_missing_count = listings.isna().sum()[listings.isna().sum() != 0]
col_missing_count.sort_values(ascending=False, inplace=True)
col_missing_count

neighborhood_overview          30259
host_since                     30259
host_neighbourhood             30259
host_verifications             30259
host_total_listings_count      30259
host_thumbnail_url             30259
host_response_time             30259
host_acceptance_rate           30259
host_response_rate             30259
calendar_updated               30259
instant_bookable               30259
neighbourhood                  30259
license                        24973
host_about                     12410
bedrooms                       10973
bathrooms                      10112
beds                            9420
price_quote_price_per_night     8745
price_quote_total_price         8745
estimated_revenue_l365d         8744
price                           8744
review_scores_value             8568
review_scores_location          8568
review_scores_checkin           8567
review_scores_accuracy          8565
review_scores_communication     8562
review_scores_cleanliness       8559
r

In [8]:
col_missing_pct = (listings.isna().mean()[listings.isna().mean() != 0]) * 100

col_missing_pct.sort_values(ascending=False, inplace=True)

col_missing_pct = round(col_missing_pct,2)
fully_empty_columns = col_missing_pct[col_missing_pct == 100]
partially_missing_columns = col_missing_pct[col_missing_pct < 100]
# Notice that we used the missing values percent per column because
# when the data increases in the futrue this won't get affected but if we
# used the total missing count (30259 now) this would change when data size increases

print("Fully empty columns:")
print(fully_empty_columns)

print("\nPartially missing columns:")
print(partially_missing_columns)

Fully empty columns:
neighborhood_overview        100.0
host_since                   100.0
host_neighbourhood           100.0
host_verifications           100.0
host_total_listings_count    100.0
host_thumbnail_url           100.0
host_response_time           100.0
host_acceptance_rate         100.0
host_response_rate           100.0
calendar_updated             100.0
instant_bookable             100.0
neighbourhood                100.0
dtype: float64

Partially missing columns:
license                        82.53
host_about                     41.01
bedrooms                       36.26
bathrooms                      33.42
beds                           31.13
price_quote_price_per_night    28.90
price_quote_total_price        28.90
estimated_revenue_l365d        28.90
price                          28.90
review_scores_value            28.32
review_scores_location         28.32
review_scores_checkin          28.31
review_scores_accuracy         28.31
review_scores_communication    28.3

In [9]:
rows_missing_count = listings.isna().any(axis=1).sum()

rows_missing_pct = listings.isna().any(axis=1).mean() * 100

print(f"Rows with at least one missing value: {rows_missing_count:,}")
print(f"Overall percentage of affected rows:  {rows_missing_pct:.2f}%")

Rows with at least one missing value: 30,259
Overall percentage of affected rows:  100.00%


# Drop unusable columns
Drop columns that are entirely empty (e.g. `host_since`, `host_response_time`, `neighbourhood`, `instant_bookable`, `calendar_updated`) — they can't be cleaned or imputed, only removed.

In [10]:
columns_before_drop = listings.columns.__len__()

In [11]:
listings.drop(columns=fully_empty_columns.index, inplace=True)

In [12]:
columns_after_drop = listings.columns.__len__()

# Clean the price column
`price` is stored as text with a leading "$" (e.g. "$113.97"). Strip the currency symbol and commas, then cast to `float` so it can be used numerically.

In [13]:
listings['price'] = (
    listings['price']
    .str.replace('$', '')
    .str.replace(',', ''))
listings.price.head(3)

0    113.97
1    117.27
2     80.06
Name: price, dtype: str

In [14]:
listings.price = listings.price.astype('float64')
listings.price.dtype

dtype('float64')

# Fill numeric gaps	
Fill missing `bedrooms`, `bathrooms`, and `beds` using the `median` value for that listing's `room_type` and `accommodates` group (not the overall `median`).

In [15]:
group_cols = ['room_type','accommodates']

numeric_cols = ['bedrooms','bathrooms','beds']

print(listings[numeric_cols].isna().sum())
print(t)
print(listings.dtypes[numeric_cols])

bedrooms     10973
bathrooms    10112
beds          9420
dtype: int64
---------------------------------------
bedrooms     float64
bathrooms    float64
beds         float64
dtype: object


In [16]:
for col in numeric_cols:
    group_median = listings.groupby(group_cols)[col].transform('median')
    listings.fillna({col:group_median}, inplace=True)
print(listings[numeric_cols].isna().sum())

bedrooms     0
bathrooms    2
beds         6
dtype: int64


In [17]:
# First-pass logic is correct. but we still have missing values. 
# For these exceptional cases, use a documented fallback: the median for the same room_type alone.
for col in numeric_cols:
    group_median = listings.groupby('room_type')[col].transform('median')
    listings.fillna({col:group_median}, inplace=True)
print(listings[numeric_cols].isna().sum())

bedrooms     0
bathrooms    0
beds         0
dtype: int64


# Handle the license column
`license` is only ~82% populated. A missing license may be meaningful (unregistered listing) rather than a data error — add a boolean `has_license` column instead of dropping or imputing rows.

In [20]:
round(listings.license.isna().mean() * 100,2)

listings['has_license'] = listings['license'].notna()

listings.has_license.unique()

array([False,  True])

# Fix data types
Convert `neighbourhood_group_cleansed` and `room_type` to `category` dtype, and confirm `id` and `host_id` are read as `integers`, not `floats` or `strings`.

In [25]:
listings[['neighbourhood_group_cleansed','room_type','id','host_id']].dtypes

neighbourhood_group_cleansed    category
room_type                       category
id                                 int64
host_id                            int64
dtype: object

In [24]:
listings[['neighbourhood_group_cleansed',
            'room_type']] = listings[
            ['neighbourhood_group_cleansed','room_type']
            ].astype('category')


In [26]:
listings[['neighbourhood_group_cleansed','room_type']].dtypes

neighbourhood_group_cleansed    category
room_type                       category
dtype: object

# Create a price-per-guest metric
Add a column price_per_accommodate = price / accommodates, using your cleaned numeric price column.	vectorized column math